src/callbacks/ltfb.cpp

```diff
@@ -34,283 +33,551 @@
-                  &recv, 1, partner_trainer, 0,
-                  El::SyncInfo<El::Device::CPU>{});
-  }
-  comm.trainer_barrier();
+    // Load model checkpoint from partner trainer
+    {
+      persist p;
+      p.set_cb_type(callback_type::model_only);
+      p.open_restart(recv_dir);
+      m.load_from_checkpoint_shared(p);
+      p.close_restart();
+    }
 
-  // Load model checkpoint from partner trainer
-  {
-    persist p;
-    p.set_cb_type(callback_type::model_only);
-    p.open_restart(recv_dir);
-    m.load_from_checkpoint_shared(p);
-    p.close_restart();
+    restore_model_weights(m, restore_weights);
   }
+private:
+  std::string ckpt_basedir_;
+};// class CheckpointFile
+
+class CheckpointBlob final
+  : public Cloneable<CheckpointBlob, LTFBCommunicationAlgorithm>
+{
+  using BaseType = Cloneable<CheckpointBlob, LTFBCommunicationAlgorithm>;
+public:
+  CheckpointBlob(std::set<std::string> const& weights_names)
+    : BaseType(weights_names)
+  {}
+  CheckpointBlob(std::set<std::string>&& weights_names)
+    : BaseType(std::move(weights_names))
+  {}
+  void exchange_models(model& m,
+                       El::Int partner_trainer,
+                       El::Int /*step*/) const final
+  {
+    auto&& comm = *m.get_comm();
+    // Keep track of weights that shouldn't be exchanged
+    std::unordered_map<std::string, std::unique_ptr<weights>> restore_weights;
+    if (this->has_weights_names()) {
+      auto const& weights_names = this->weights_names();
+      for (auto w : m.get_weights()) {
+        if (weights_names.find(w->get_name()) == weights_names.cend())
+        {
+          using TensorDataType = DataType;
+          using WeightsType = data_type_weights<TensorDataType>;
+          restore_weights[w->get_name()]
+            = make_unique<WeightsType>(dynamic_cast<WeightsType&>(*w));
+        }
+      }
+    }
 
-  // Restore weights that shouldn't be exchanged
-  if (!restore_weights.empty()) {
-    for (auto w : m.get_weights()) {
-      if (restore_weights.count(w->get_name()) > 0) {
-        using TensorDataType = DataType;
-        using WeightsType = data_type_weights<TensorDataType>;
-        dynamic_cast<WeightsType&>(*w)
-          = dynamic_cast<WeightsType&>(*restore_weights[w->get_name()]);
+    // Save model checkpoint
+    std::ostringstream oss;
+    {
+      RootedBinaryOutputArchive ar(oss, comm.get_trainer_grid());
+      comm.trainer_barrier();
+      ar(m);
+    }
+
+    // sure, why not
+    comm.trainer_barrier();
+
+    // Synchronize with partner trainer
+    std::string save_model_ckpt = oss.str(), load_model_ckpt;
+    if (comm.am_trainer_master())
+    {
+      std::size_t save_size=save_model_ckpt.size(), load_size=0;
+      comm.sendrecv(&save_size, 1, partner_trainer, 0,
+                    &load_size, 1, partner_trainer, 0,
+                    El::SyncInfo<El::Device::CPU>{});
+      load_model_ckpt.resize(load_size);
+
+      while (save_size && load_size)
```

Comment:
We should handle the case where the send and load models require a different number of data blocks:
```suggestion
      while (save_size || load_size)
```

Reply:
That was my intent; brain fail. Thanks!